### PR TITLE
 Laboratory work #2, Anna Evgrafova - 23FPL1

### PR DESCRIPTION
**Pull request by Evgraff2005**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,6 +18,22 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+
+    if any(symb.isalpha() for symb in text) is False:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:24]: Is this check really necessary?

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -18,6 +18,22 @@ def tokenize(text: str) -> list[str] | None:
...
+        return []
+    new_text = []
+    word_for_new_text = []
+
+    for every_symbol in text.lower():
+        if every_symbol.isalpha():
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:30]: use continue

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -18,6 +18,22 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+
+    if any(symb.isalpha() for symb in text) is False:
+        return []
+    new_text = []
+    word_for_new_text = []
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:27]: Is it really necessary to create two lists?

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great job! Raise score and keep working!

File: lab_2_retrieval_w_bm25/main.py (hunk 2 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:30]: where?.. Please, explain

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:24]: wow, I was sure the code won't work without it but it works. Done

File: lab_2_retrieval_w_bm25/main.py (hunk 2 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:30]: in the 'for construction' there are two brunches that cannot be replaced with 'continue', none of them

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:27]: yes?.. How can I do two different actions in ine moment? In my code, I at first put separated letters into the list 'word_for_new_text', when there appears a space (the border between words) I go to the other step. In that step I make these letters to become a word using .join(). After all I place a word into the list 'new_text'.

Now I see the only way to delete 'new_text': to copy the initial one and to clear one of them but it doesn't make any sense

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -33,6 +49,20 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+    if (not isinstance(tokens, list)
+            or not all(isinstance(token, str) for token in tokens)):
+        return None
+    if (not isinstance(stopwords, list)
+            or not all(isinstance(word, str) for word in stopwords)):
+        return None
+    if len(tokens) == 0 or len(stopwords) == 0:
+        return None
+
+    new_list_of_tokens = []
```
[the remark this answers is not in the document]
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:61]: done!

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 2 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py]: now seems to be done

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -33,6 +48,19 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+    if (not isinstance(tokens, list)
+            or not all(isinstance(token, str) for token in tokens)):
+        return None
+    if (not isinstance(stopwords, list)
+            or not all(isinstance(word, str) for word in stopwords)):
+        return None
+    if len(tokens) == 0 or len(stopwords) == 0:
+        return None
+
+    document_tokens = [word for word in tokens if word not in stopwords]
```
[the remark this answers is not in the document]
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:60]: turned out I didn't listened to the lecturer for nothing (•‿•)
thanks!

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -47,6 +75,17 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+            or not all(isinstance(list_of_tokens, list) for list_of_tokens in documents)):
+        return None
+    for list_of_tokens in documents:
+        if not all(isinstance(token, str) for token in list_of_tokens):
+            return None
+    words_from_all_docs_list = list(set(word for sublist in documents for word in sublist))
+
+    if not words_from_all_docs_list:
+        return None
+    return list(words_from_all_docs_list)
```
[the remark this answers is not in the document]
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:88]: done

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -18,6 +20,19 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+
+    only_letters_text = ''
+    for letter in text:
+        if letter.isalpha() or letter == ' ':
+            only_letters_text += letter.lower()
```
[the remark this answers is not in the document]
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:29]: done

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -77,6 +137,33 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+            or not documents):
+        return None
+    for one_doc_text in documents:
+        if (not all(isinstance(word, str) for word in one_doc_text)
+                or not one_doc_text):
+            return None
+
+    idf_figure_dict = {}
+    for word in vocab:
+        in_how_many_docs_is_met = 0
```
[the remark this answers is not in the document]
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py]: Maybe I can use list comprehensions. I should create an empty list right after the beginning of the cycle. Then I need to write smth like that:

**in_how_many_docs_is_met = sum([1 for doc in documents if word in one_doc_text])**

File: lab_2_retrieval_w_bm25/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py]: It works! Hope, this version is final

File: lab_2_retrieval_w_bm25/main.py [hunk 9]
```diff
@@ -18,6 +21,19 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+
+    only_letters_text = ''
+    for letter in text.lower():
+        if letter.isalpha() or letter == ' ':
+            only_letters_text += letter
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:30]: try not to create a new string to save letters from the text, but use `replace` in current text

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Vasilisa-Blyudova) [reply lab_2_retrieval_w_bm25/main.py:61]: I am very pleased!

File: lab_2_retrieval_w_bm25/main.py [hunk 10]
```diff
@@ -47,6 +76,17 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if (not isinstance(documents, list)
+            or not all(isinstance(list_of_tokens, list) for list_of_tokens in documents)):
+        return None
+    for list_of_tokens in documents:
+        if not all(isinstance(token, str) for token in list_of_tokens):
+            return None
+    words_from_all_docs_list = list(set(word for sublist in documents for word in sublist))
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:85]: very good!

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -77,6 +138,31 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+            or not documents):
+        return None
+    for one_doc_text in documents:
+        if (not all(isinstance(word, str) for word in one_doc_text)
+                or not one_doc_text):
+            return None
+
+    idf_figure_dict = {}
+    for word in vocab:
+        for one_doc_text in documents:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:156]: Do you really need to duplicate this condition? It's already on the next line

File: lab_2_retrieval_w_bm25/main.py [hunk 12]
```diff
@@ -138,6 +282,39 @@ def rank_documents(
...
+    tokenized_query_without_stopwords = remove_stopwords(tokenized_query, stopwords)
+    if not isinstance(tokenized_query_without_stopwords, list):
+        return None
+    l_of_tuples_dict_plus_its_relevance = []
+
+    for dict_index, dict_at_the_index in enumerate(indexes):
+        relevance_of_the_particular_doc = 0.0
+        for word in tokenized_query_without_stopwords:
+            if word in dict_at_the_index:
+                relevance_of_the_particular_doc += dict_at_the_index[word]
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:308]: can we also use sum() as in `calculate_idf` function?

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Very good work! All you have to do is fix the comments!

File: lab_2_retrieval_w_bm25/main.py (hunk 11 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:156]: image
this variable will not work(

File: lab_2_retrieval_w_bm25/main.py (hunk 9 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:30]: done!

File: lab_2_retrieval_w_bm25/main.py (hunk 12 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py]: done!

File: lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json [hunk 13]
```diff
@@ -36,4 +36,4 @@
         "output": 0.0,
         "across": 0.3664
     }
-]
+]
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json:39]: revert

File: lab_2_retrieval_w_bm25/main.py [hunk 14]
```diff
@@ -5,6 +5,9 @@
 """
 # pylint:disable=too-many-arguments, unused-argument
 
+import json
+import math
+
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:10]: rm

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Very good work! All you have to do is fix the comments!

File: lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json (hunk 13 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json:39]: what?.. I don't understand

Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json:39]: please, explain

File: lab_2_retrieval_w_bm25/main.py (hunk 14 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:10]: what should I remove?

File: lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json (hunk 13 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json:39]: done!

File: lab_2_retrieval_w_bm25/main.py (hunk 14 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:10]: done! (I removed the 7th line)

File: lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json [hunk 15]
```diff
@@ -36,4 +36,4 @@
         "output": 0.0,
         "across": 0.3664
     }
-]
\ No newline at end of file
+]
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json:39]: revert

File: lab_2_retrieval_w_bm25/main.py [hunk 16]
```diff
@@ -120,6 +222,45 @@ def calculate_bm25(
...
+            return None
+        if word_from_idf not in bm25_dict:
+            bm25_dict[word_from_idf] = 0.0
+        freq = document.count(word_from_idf)
+        bm25_figure = (idf_document[word_from_idf]
+                       * ((freq * (k1 + 1)) / (freq + k1
+                                               * (1 - b + b * (doc_len / avg_doc_len)))))
+        bm25_dict[word_from_idf] = bm25_figure
+
+    if not bm25_dict:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:261]: return bm25_dict or None

File: lab_2_retrieval_w_bm25/main.py [hunk 17]
```diff
@@ -138,6 +279,39 @@ def rank_documents(
...
+        for word in tokenized_query_without_stopwords:
+            if word in dict_at_the_index:
+                relevance_of_the_particular_doc.append(dict_at_the_index[word])
+        l_of_tuples_dict_plus_its_relevance.append((dict_index,
+                                                    sum(relevance_of_the_particular_doc)))
+
+    l_of_tuples_dict_plus_its_relevance = sorted(l_of_tuples_dict_plus_its_relevance,
+                                                 key=lambda x: x[1], reverse=True)
+
+    if not l_of_tuples_dict_plus_its_relevance:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:312]: .

File: lab_2_retrieval_w_bm25/main.py [hunk 18]
```diff
@@ -92,6 +175,25 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
...
+        return None
+
+    tf_idf_dict = {}
+    for tf_word in tf.keys():
+        if tf_word not in idf.keys():
+            return None
+        tf_idf_result = tf[tf_word] * idf[tf_word]
+        tf_idf_dict[tf_word] = tf_idf_result
+
+    if not tf_idf_dict:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:194]: .

File: lab_2_retrieval_w_bm25/main.py [hunk 19]
```diff
@@ -33,6 +46,19 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+            or not all(isinstance(token, str) for token in tokens)):
+        return None
+    if (not isinstance(stopwords, list)
+            or not all(isinstance(word, str) for word in stopwords)):
+        return None
+    if len(tokens) == 0 or len(stopwords) == 0:
+        return None
+
+    document_tokens = [word for word in tokens if word not in stopwords]
+    if not document_tokens:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:59]: .

**Review by Vasilisa-Blyudova — CHANGES_REQUESTED** [review]

File: lab_2_retrieval_w_bm25/main.py (hunk 16 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:261]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 17 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:312]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 18 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:194]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 19 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/main.py:59]: done

File: lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json (hunk 15 above)
Comment (Evgraff2005) [reply lab_2_retrieval_w_bm25/tests/assets/saved_metrics_example.json]: the mentor says everything is alright now, there is nothing to revert. So, I suppose, done

**Review by Vasilisa-Blyudova — DISMISSED** [review]

**Pull request closed without merging**